### DISCUSSION TITLE
gRPC .proto imports not possible?

### DISCUSSION THREAD

**Discussion by ghost** [microcks/microcks#830] [Q&A] [answered]
Hi,

I want to import other .proto files in my main gRPC service .proto file.

If I'll use `import "google/protobuf/empty.proto";` it works.

If i'll use custom .proto file like `import "custom.proto" ` in my gRPC service proto file, microcks tells me a parsing error.

I also didn't find an option to actually import the needed proto files in beforehand.

Maybe it is not yet implemented and related to this issue: [https://github.com/microcks/microcks/issues/449](https://github.com/microcks/microcks/issues/449)

**Comment by lbroudoux** [comment]
Hi @lennakai 

custom gRPC proto resolution should be done when importing using a URL as in a scheduled import (standard one as resolved within embedded gRPC libs). Are you suing this method for importing your proto?

**Reply by ghost** [reply]
No, I will try this out. 
Before I used to import them manually and then merged everything together before I used the scheduled import.

**Reply by ghost** [reply]
@lbroudoux
I tried it out, but it didn't work.

I created a small [repo](https://github.com/lennakai/microcks-test/blob/main/) from where I did my test.

For my .proto scheduled importer it shows "Import Error"

**Reply by lbroudoux** [reply]
Hum... Thank you! Will have a look at it.

**Reply by lbroudoux** [reply]
Wondering if we don't have a bug regarding `google/protobuf` lib import... Also are you sure `int` is a valid type in `custom.proto`? I use to see `int32` or `int64` instead... (but not a protobuf expert)

**Reply by ghost** [reply]
I think you are right about the `int`. I changed it accordingly. Still didn't work for me. 

Maybe i am doing something wrong in the way I import it.

**Reply by lbroudoux** [reply, marked as answer]
Nope! You do it right. I just find out that we have a bug...

When doing direct upload import, `google/protobuf` is locally resolved but the import fails on `custom.proto` not found.
When doing scheduled download import, `custom.proto` is correctly found and compiles correctly (once `int` changed to `int32`) but we try to resolve `google/protobuf` remotely and it is not present in your repo 😉 

A similar unit test I have is actually using Google libs that are put into the repo as well. See https://github.com/microcks/microcks/blob/master/webapp/src/test/resources/io/github/microcks/util/grpc/hello-v1-option.proto

So, as of today, things will work if you also put `google/protobuf/empty.proto` into your repo. However, I think we can assume to rely on the `google/protobuf/*.proto` files that are embedded in compiler and avoid trying to download the one from your repo. What do you think?

**Reply by ghost** [reply]
Yes, I think so too. But it would mean, that if google changes these files in the future, we always have the newest version in microcks?

Thank you for your help.

**Reply by lbroudoux** [reply]
We provide and maintain them within the `protoc-jar` module (see https://github.com/microcks/protoc-jar/blob/master/src/main/resources/proto3/include/google/protobuf/empty.proto). The thing I'm not aware of is how Google is committed to not change those files during the lifecycle of a compiler version... this is necessary to avoid unexpected behavior or compilation issues. Will have a look at it.

**Reply by lbroudoux** [reply]
This has been fixed here: https://github.com/microcks/microcks/issues/832
